nixos/roles/slurm/fc-set-munge-key.py

```diff
@@ -0,0 +1,77 @@
+import filecmp
+import hashlib
+import json
+import os
+import shutil
+import sys
+from pathlib import Path
+
+import pystemd.systemd1
+from rich import print
+
+required_env_vars = {"encServicesPath", "mungeKeyFile"}
+missing_env_vars = required_env_vars - os.environ.keys()
+
+os.umask(0o0266)
+
+if missing_env_vars:
+    print(f"error: missing environment variables: {missing_env_vars}")
+    sys.exit(1)
+
+enc_services_path = Path(os.environ["encServicesPath"])
+if not enc_services_path.exists():
+    print(f"error: services files at {enc_services_path} is missing")
+    sys.exit(2)
+
+munge_key_path = Path(os.environ["mungeKeyFile"])
+
+with open(enc_services_path) as f:
+    enc_services = json.load(f)
+
+controller_service = [
+    s for s in enc_services if s["service"] == "slurm-controller-controller"
+]
+
+if not controller_service:
+    print("error: service not found, is the slurm-controller role enabled?")
+    sys.exit(3)
+
+service_pw = controller_service[0]["password"] + "\n"
+munge_key = hashlib.sha256(service_pw.encode()).hexdigest()[:64]
+
+key_dir = munge_key_path.parent
+if not key_dir.exists():
+    print(f"init: creating key dir at {key_dir}")
+    key_dir.mkdir()
+
+tmp_key = key_dir / "tmp-set-munge-key"
```

Comment:
We still always use the same file name – this time of the tmpfile. So there could still be the very slight possibility of both processes having opened the same file concurrently and thus writing garbage.

Or do we rely on systemd here for ensuring only a single service run concurrently?

Reply:
I trust systemd here :)